packages/otel/src/trace-effect.ts

```diff
@@ -0,0 +1,123 @@
+
+const TraceEffectAttributes = {
+  DURATION_MS: "trace_effect.duration_ms",
+  SLOW_THRESHOLD_MS: "trace_effect.slow_threshold_ms",
+  IS_SLOW: "trace_effect.is_slow",
+} as const;
+
+export interface TraceEffectLogContext {
+  attributes: Record<string, unknown>;
+  durationMs?: number;
+  slowThresholdMs?: number;
+  error?: string;
+}
+
+export interface TraceEffectCallbacks {
+  onStart?: (name: string, ctx: TraceEffectLogContext) => void;
+  onFinish?: (name: string, ctx: TraceEffectLogContext) => void;
+  onSlow?: (name: string, ctx: TraceEffectLogContext) => void;
+  onError?: (name: string, ctx: TraceEffectLogContext) => void;
+}
+
+export interface TraceEffectOptions {
+  name: string;
+  slowThresholdMs?: number;
+  callbacks?: TraceEffectCallbacks;
+}
+
+/**
+ * Creates a tracer function that wraps async operations with OTEL spans.
+ * Creates a span for each operation and optionally calls logging callbacks.
+ * Useful for tracing external calls (APIs, databases) to investigate timeouts.
+ *
+ * @example
+ * // Basic usage (spans only, no logging)
+ * const traceAlgolia = createTraceEffect({ name: "Algolia saveObjects", slowThresholdMs: 10000 });
+ * await traceAlgolia(() => index.saveObjects(objects), { indexName, count: objects.length });
+ *
+ * @example
+ * // With logging callbacks
+ * const traceAlgolia = createTraceEffect({
+ *   name: "Algolia saveObjects",
+ *   slowThresholdMs: 10000,
+ *   callbacks: {
+ *     onStart: (name, ctx) => logger.debug(`Starting: ${name}`, ctx.attributes),
+ *     onFinish: (name, ctx) => logger.debug(`Finished: ${name}`, ctx),
+ *     onSlow: (name, ctx) => logger.warn(`Slow: ${name}`, ctx),
+ *     onError: (name, ctx) => logger.error(`Failed: ${name}`, ctx),
+ *   },
+ * });
+ */
+export function createTraceEffect(options: TraceEffectOptions) {
+  const { name, slowThresholdMs = DEFAULT_SLOW_THRESHOLD_MS, callbacks } = options;
+
+  return async function traceEffect<T>(
+    operation: () => Promise<T>,
+    attributes: Record<string, unknown> = {},
+  ): Promise<T> {
+    const span = tracer.startSpan(name, {}, context.active());
+
+    // Set initial attributes on span
+    Object.entries(attributes).forEach(([key, value]) => {
+      if (value !== undefined && value !== null) {
+        span.setAttribute(key, String(value));
+      }
+    });
+
+    callbacks?.onStart?.(name, { attributes });
+
+    const startTime = performance.now();
+
+    try {
+      const result = await operation();
+      const durationMs = Math.round(performance.now() - startTime);
+
+      span.setAttribute(TraceEffectAttributes.DURATION_MS, durationMs);
+      span.setAttribute(TraceEffectAttributes.SLOW_THRESHOLD_MS, slowThresholdMs);
+
+      if (durationMs > slowThresholdMs) {
+        span.setAttribute(TraceEffectAttributes.IS_SLOW, true);
+        span.setStatus({
+          code: SpanStatusCode.ERROR,
+          message: `Operation exceeded slow threshold: ${durationMs}ms > ${slowThresholdMs}ms`,
```

Comment:
I have seconds thoughs. We can merge it and check but Im worried we will produce false alarms - without context, slow span itself is really slow 🤔

we can merge it as mvp but maybe long term we should add custom attribute "SLOW: true" or somthing and add custom views in datadog. or maybe we should add metrics?